src/NuGet.Clients/NuGet.Credentials/CredentialService.cs

```diff
@@ -0,0 +1,191 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Credentials
+{
+    /// <summary>
+    /// This service manages orchestrates credential providers and supplies credentials
+    /// for use in http requests
+    /// </summary>
+    public class CredentialService : ICredentialService
+    {
+        private static IEnumerable<ICredentialProvider> _defaultProviders
+            = new List<ICredentialProvider>();
+
+        private IEnumerable<ICredentialProvider> _providerList = null;
+
+        private readonly ConcurrentDictionary<string, bool> _retryCache
+            = new ConcurrentDictionary<string, bool>();
+        private readonly ConcurrentDictionary<string, ICredentials> _providerCredentialCache
+            = new ConcurrentDictionary<string, ICredentials>();
+
+        private readonly bool _nonInteractive;
+        private readonly bool _useCache;
+        /// <summary>
+        /// This semaphore ensures only one provider active per process, in order
+        /// to prevent multiple concurrent interactive login dialogues.
+        /// Unnamed semaphores are local to the current process.
+        /// </summary>
+        private readonly Semaphore _providerSemaphore = new Semaphore(1, 1);
```

Comment:
This should be a SemaphoreSlim so that .WaitAsync can be used in the async method below